cuda_phase-retrieval/phase_algo.hpp

```diff
@@ -0,0 +1,419 @@
+{
+    cufftDoubleComplex complex1i, exp_target;
+    complex1i.x = 0; complex1i.y = 1;
+    for (int idx = blockIdx.x * blockDim.x + threadIdx.x; idx < dimension; idx += blockDim.x * gridDim.x)  
+    {
+        //arg = atan2(imag, real)
+        exp_target.x = atan2(x_hat[idx].y, x_hat[idx].x);
+        exp_target.y = 0;
+        y_hat[idx] = cuCmul(ptrMag[idx], gpu_exp(cuCmul(complex1i, exp_target)));
+    }
+}
+
+//processing magnitudes with mask
+__global__ void process_arrays(double *mask, cufftDoubleComplex *y_hat, double *image_x, double *image_x_p, cufftDoubleComplex *image_x_comp, double beta, int mode, int iter, int dimension)
+{
+    for (int idx = blockIdx.x * blockDim.x + threadIdx.x; idx < dimension; idx += blockDim.x * gridDim.x)  
+    {
+        bool logical_not_mask;
+        bool y_less_than_zero;
+        bool logical_and;
+        bool indices;
+
+        double y = get_real(normalize(y_hat[idx], dimension));
+
+        if(iter == 0) image_x_p[idx] = y;
+        else image_x_p[idx] = image_x[idx];
+
+        //updates for elements that satisfy object domain constraints
+        if(mode == 3 || mode == 1) image_x[idx] = y;
+
+        //find elements that violate object domain constraints or are not masked 
+        if(mask[idx] <= 0) logical_not_mask = true;
+        else if(mask[idx] >= 1) logical_not_mask = false;
+
+        //check if any element y is less than zero
+        if(y < 0) y_less_than_zero = true;
+        else if(y >= 0) y_less_than_zero = false;
+
+        //use "and" logical to check the "less than zero y" and the mask  
+        if(y_less_than_zero == true && mask[idx] >= 1) logical_and = true;
+        else logical_and = false;
+
+        //create indices with logical "not"
+        if(logical_and == false && logical_not_mask == false) indices = false;
+        else indices = true;
+
+        //updates for elements that violate object domain constraints
+        if(indices == true)
+        {
+            if(mode == 1 || mode == 2)
+            {
+                image_x[idx] = image_x_p[idx]-beta*y;
+            }
+            else if(mode == 3)
+            {
+                image_x[idx] = y-beta*y;
+            }
+        }
+
+        image_x_comp[idx] = get_complex(image_x[idx]);
+    }
+}
+
+//CUFFT error checking
+void CUFFT_CHECK(cufftResult cufft_process)
+{
+    if(cufft_process != CUFFT_SUCCESS) cout<<cufft_process<<endl;
+}
+
+//Create a folder
+void createDir(string dir)
+{
+    #if defined _MSC_VER
+        _mkdir(dir.data());
+    #else
+        mkdir(dir.data(), 0777);
+    #endif
+}
+
+//Get current location (same as the executable file)
+string get_current_dir() 
+{
+   char buff[FILENAME_MAX]; //create string buffer to hold path
+   GetCurrentDir( buff, FILENAME_MAX );
+   string current_working_dir(buff);
+   return current_working_dir;
+}
+
+//Create grayscale image file in a folder
+void print_image(string workdir, vector<double> image, int iter, int size_x, int size_y)
```

Comment:
OpenCV should only used in the test cases, because it is a heavy dependency and does not add new functionality to the actual algorithm. Maybe we should make OpenCV also optional in the test cases.